### PR TITLE
Testing Nickel snippets in the manual

### LINKED ISSUE #1154: Check evaluation of snippets in the manual
During our quest to rename the stdlib I hit a pain point with keeping the manual up-to-date. It would have been very helpful if we had some automated way to check that the Nickel snippets in it are valid.

I'm imagining some tooling that extracts code snippets marked as `nickel` from the manual and tried to evaluate them. This could then be part of the CI. There are probably lots of edge cases with incomplete code fragments and intentionally failing programs, so this might get a bit tricky. But maybe it's still worth the effort.

### PR DESCRIPTION
**Pull request by vkleen**
Add a new test suite in `core/tests/manual` which parses the Markdown files in `doc/manual` to extract all Nickel code blocks and checks that they parse or evaluate correctly. For now, REPL sessions in the manual aren't evaluated during testing but only parsed.

To facilitate this, code blocks in the manual are annotated with `nickel`, `nickel-lines`, `nickel-parse`, `nickel-repl` or `nickel-no-check` depending on the parsing or evaluation mode desired.

Closes #1154

### CODE REVIEW

File: doc/manual/contracts.md [hunk 1]
```diff
@@ -19,13 +19,16 @@ using the `|` operator:
 let x = (1 + 1 | Number) in x
 ```
 
-Contracts can also be attached to identifiers in a definition:
+Contracts can also be attached to identifiers in a definition. In let bindings,
+the following is equivalent to the previous example:
```
Comment (jneem) [inline doc/manual/contracts.md:23]:
```suggestion
Contracts can also be attached to identifiers in a let binding.
The following is equivalent to the previous example:
```

File: doc/manual/contracts.md [hunk 2]
```diff
@@ -100,16 +103,16 @@ In `IsFoo`, we first test if the value is a string, and then if it is equal to
 `"foo"`, in which case the contract succeeds. Otherwise, the contract fails with
 appropriate error messages. Let us try:
 
-```nickel
+```nickel-repl
 nickel> 1 | IsFoo
```
Comment (jneem) [inline doc/manual/contracts.md:107]:
```suggestion
> 1 | IsFoo
```

File: doc/manual/syntax.md [hunk 3]
```diff
@@ -147,13 +147,13 @@ present on all lines of the string is stripped. This way, multiline strings can
...
 
-```nickel
+```nickel-repl
 > m%"
-  This line has no indentation.
-    This line is indented.
-      This line is even more indented.
-  This line has no more indentation.
-"%
+    This line has no indentation.
```
Comment (jneem) [inline doc/manual/syntax.md:152]: I'm a little confused by this example. I guess you had to indent it extra so that the final `"%` will be caught by the parser? But even before this change, the "This line has no indentation" line had indentation...

File: doc/manual/typing.md [hunk 4]
```diff
@@ -290,25 +294,23 @@ let snd : forall a b. a -> b -> b = fun x y => y in
...
-<!-- ``` -->
+Or even nest them:
+
+```nickel
+let higherRankId : forall a. (forall b. b -> b) -> a -> a
+  = fun id x => id x in
+let id : forall a. a -> a
+  = fun x => x in
+higherRankId id 0 : Number
+```
```
Comment (jneem) [inline doc/manual/typing.md:305]: :tada:

**Review by jneem — APPROVED** [review]

File: doc/manual/syntax.md (hunk 3 above)
Comment (vkleen) [reply doc/manual/syntax.md:152]: The multiline string parser strips common indentation. So this line doesn't have any indentation in the final string and that's what this example is supposed to illustrate.

File: doc/manual/typing.md [hunk 5]
```diff
@@ -676,6 +678,13 @@ can be sure that something will eventually be a port number, or will fail with
 the correct error message: when using a contract application.
 
 ```nickel
+# hide-start
+let Port = std.contract.from_predicate (fun value =>
+  std.is_number value
+  && value % 1 == 0
```
[the remark this answers is not in the document]
Comment (vkleen) [reply doc/manual/typing.md:684]: If `value` isn't an integer, `value % 1` returns the fractional part of `value`. I suppose that's in line with writing `x mod 1` for the residue class of `x` in $\mathbb R/\mathbb Z$...

File: core/tests/manual/main.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+enum CodeBlockType {
+    /// A code block intended to be evaluated in its entirety
+    Single,
+    /// A code block containing multiple lines, each of which should be
+    /// evaluated separately
+    Lines,
+    /// A code block that should only be parsed, not evaluated
+    Parse,
+    /// A code block containing a REPL session, e.g.
+    /// ```nickel repl
```
Comment (yannham) [inline core/tests/manual/main.rs:22]: The example doesn't check out with practice: it seems that you are requiring a `-` instead of a space, ie `nickel-repl`.

File: doc/manual/contracts.md [hunk 7]
```diff
@@ -711,34 +717,37 @@ value and continues with the second argument (here, our wrapped `value`).
...
-  "0" | doc "Some information" = true,
-}
-nickel>:query config "0"
+```nickel-repl
+> let config | NumberBoolDict = {
+    "1" | std.FailWith "ooch" = null, # same as our previous "fail"
+    "0" | doc "Some information" = true,
+  }
+
+# Don't parse this in tests hide-line
```
Comment (yannham) [inline doc/manual/contracts.md:726]: What's the purpose of this comment, and where is it handled? I think I missed it somehow... Does it appear in the final manual? Also, I hope we can get rid of it eventually - starting an REPL engine using `ReplImpl` to actually test the output of the REPL examples should be doable.

**Review by yannham — COMMENTED** [review]
A good step forward!

I wonder if putting a space after `nickel` in the textinfo of fenced code block isn't more idiomatic markdown. In particular, according to the common mark spec, only the first word of the textinfo (up to the first space) is considered for highlighting purpose (see eg https://spec.commonmark.org/0.28/#example-112). We adapted the website, but for example I often read the manual inside NeoVim which has syntax coloring for Nickel snippet, and who knows what markdown rendered people will use to read the manual outside of the website. Would that make sense to do:

- `nickel-repl` => `nickel repl`
- `nickel-parse` => `nickel parse`

It could be something even more specific to indicate that it doesn't concern rendering, such as `nickel @test.parse-only` (this is an absolutely random suggestion, not necessarily a good one)

File: doc/manual/contracts.md (hunk 7 above)
Comment (vkleen) [reply doc/manual/contracts.md:726]: It's a bit convoluted. The code parsing out REPL inputs expects to find a `>` directly following an empty line, so it will skip this particular input because of the comment. The `hide-line` is for gatsby to not display the comment on the website.

File: core/tests/manual/main.rs (hunk 6 above)
Comment (vkleen) [reply core/tests/manual/main.rs:22]: Right, I tried what you were suggesting about spaces before. Unfortunately, gatsby doesn't seem to ignore the extra information after `nickel`. In fact, it seems to remove the space after `nickel` and try to look for a Prismjs language `nickelrepl` if given `nickel repl` in the code block. Relying on that behaviour felt a bit too spooky for me, though...

Comment (vkleen) [reply core/tests/manual/main.rs:22]: On the other hand, relying on it and just continuing to check when updating the website should make syntax highlighting work in saner markdown renderers, I guess. It's easy to change back here, so whichever you think would make more sense long term.

File: doc/manual/contracts.md (hunk 7 above)
Comment (yannham) [reply doc/manual/contracts.md:726]: Ah. I would have suggested to add a comment to explain that, but...that's probably a bad idea :stuck_out_tongue: is the `hide-line` a vanilla Gatsby feature or something we filter out specifically in nickel-lang.org ?

Comment (vkleen) [reply doc/manual/contracts.md:726]: The `hide-line` is a feature of the [`gatsby-remark-prismjs` plugin](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-remark-prismjs/README.md#line-hiding).

File: core/tests/manual/main.rs (hunk 6 above)
Comment (yannham) [reply core/tests/manual/main.rs:22]: Yeah, and what about using a special character and a space, something like `nickel #repl` or `nickel !repl`? Gatsby would parse that as `nickel#repl` which has almost no chance of conflicting with anything, keeping the alias approach (could even be `nickel -repl`, even if that feels...slightly wrong for some reason), and such that other markdown parser would be ok with it? I tried it into Neovim and some random HTML markdown renderer extension I have and indeed adding a space preserves highlighting, while the dash doesn't. It seems like it doesn't make a difference for Gatsby in one direction or the other, but preserve other renderers

**Review by yannham — APPROVED** [review]
Whatever approach you choose for the textinfo string, it's not a blocker.

File: core/tests/manual/main.rs (hunk 6 above)
Comment (vkleen) [reply core/tests/manual/main.rs:22]: I like `nickel #...` 👍 I'll make the change here and update the gatsby config on the website.
